jina/types/document/__init__.py

```diff
@@ -354,120 +355,73 @@ def tags(self, value: Dict):
+        for field_name in self._pb_body.DESCRIPTOR.fields_by_name.keys():
+            try:
+                # ``HasField`` Checks if a certain field is set for the message.
+                if destination._pb_body.HasField(
+                    field_name
+                ) and not source._pb_body.HasField(field_name):
+                    fields_to_preserve.append(field_name)
+            except ValueError:
+                continue  # pass on purpose, HasField() only works for message fields.
+        dest_copy = deepcopy(destination)
+        if fields:
+            # Field masks are used to specify a subset of fields that should be
+            # returned by a get operation or modified by an update operation.
+            # Not using FieldMask since tags as `StructView` can not call `MergeFrom`.
+            for field in fields:
+                destination._pb_body.ClearField(field)
+                setattr(destination, field, getattr(source, field))
+        else:
+            # Merge fields completely while keeping the preserved fields.
             destination.CopyFrom(source)
-        elif include_fields is not None and exclude_fields is None:
-            FieldMask(paths=include_fields).MergeMessage(
-                source.proto,
-                destination.proto,
-                replace_message_field=replace_message_field,
-                replace_repeated_field=replace_repeated_field,
-            )
-        elif exclude_fields is not None:
-            empty_doc = jina_pb2.DocumentProto()
-
-            _dest = jina_pb2.DocumentProto()
-            # backup exclude fields in destination
-            FieldMask(paths=exclude_fields).MergeMessage(
-                destination.proto,
-                _dest,
-                replace_repeated_field=True,
-                replace_message_field=True,
-            )
-
-            if include_fields is None:
-                # override dest with src
-                destination.CopyFrom(source)
-            else:
-                # only update include fields
-                FieldMask(paths=include_fields).MergeMessage(
-                    source.proto,
-                    destination.proto,
-                    replace_message_field=replace_message_field,
-                    replace_repeated_field=replace_repeated_field,
-                )
-
-            # clear the exclude fields
-            FieldMask(paths=exclude_fields).MergeMessage(
-                empty_doc,
-                destination.proto,
-                replace_repeated_field=True,
-                replace_message_field=True,
-            )
-
-            # recover exclude fields
-            destination.proto.MergeFrom(_dest)
+            if fields_to_preserve:
+                for field in fields_to_preserve:
+                    setattr(destination, field, getattr(dest_copy, field))
+        # For the tags, stay consistent with the python update method.
+        self._pb_body.tags.update(dest_copy.tags)
```

Comment:
so all other fields are overridden during the update, but `tags` are expanded? I'm not really convinced by this inconsistency.

Reply:
true, I'll think about a better solution

Reply:
@nan-wang how do you think? shall we replace the tags or expand it?

Reply:
> so all other fields are overridden during the update
@hanxiao During `update`, the existing fields are overridden and the non-existing fields are added. The same logic goes for `tags` because `tags` is a dictionary by itself. The existing key/value pairs in the tags are overridden and only the non-existing ones are added.